packages/atomic/cypress/integration/common-actions.ts

```diff
@@ -0,0 +1,53 @@
+function observeRecursively(
+  element: Node,
+  callback: (mutations: MutationRecord[]) => void,
+  isAlreadyObserved = false
+): () => void {
+  const disconnectCallbacks: (() => void)[] = [];
+  if (!isAlreadyObserved) {
+    const observer = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        Array.from(mutation.addedNodes).forEach((node) => {
+          disconnectCallbacks.push(observeRecursively(node, callback, true));
+        });
+        callback(mutations);
+      });
+    });
+    observer.observe(element, {
+      subtree: true,
+      childList: true,
+      attributes: true,
+      characterData: true,
+    });
+    disconnectCallbacks.push(() => observer.disconnect());
+  }
+  if ((element as HTMLElement).shadowRoot) {
+    disconnectCallbacks.push(
+      observeRecursively((element as HTMLElement).shadowRoot!, callback, false)
+    );
+  }
+  Array.from(element.childNodes).forEach((child) =>
+    disconnectCallbacks.push(observeRecursively(child, callback, true))
+  );
+  return () => disconnectCallbacks.forEach((disconnect) => disconnect());
+}
+
+export function waitUntilNextSearchRendered() {
```

Comment:
Before going down this path, have we explored using cypress's [built-in retry-ability](https://docs.cypress.io/guides/core-concepts/retry-ability#Commands-vs-assertions) mechanism?

If we expect at a certain point in a test that there will be e.g. two active checkboxes, we can query for them and assert that it is the case.

```
cy.get('.active-checkbox-class')
    .should('have.length', 2)
```

If the assertion fails, cypress will retry until it reaches the timeout. If the assertion passes, then we can proceed with the next step in test.

Could adding these intermediate check-points can avoid the mutation observers?

Reply:
Good point. I will try instead fixing errors on a case-by-case basis with assertions.

However, concerning retry-ability, we use a lot of `before()` calls, and [those cannot be retried](https://docs.cypress.io/guides/guides/test-retries#:~:text=However%2C%20failures%20in%20before%20and%20after%20hooks%20will%20not%20trigger%20a%20retry.).